app/model/repository/Groups.php

```diff
@@ -103,13 +106,86 @@ public function findByName(string $locale, string $name, Instance $instance, ?Gr
+   *  are returned. Otherwise, all groups are returned.
+   * @param string|null $instanceId ID of an instance to which the groups belongs to.
+   * @param string|null $search Search query.
+   * @param bool|null $archived Whether to include archived groups.
+   * @param bool|null $onlyArchived Automatically implies $archived flag and returns only archived groups.
+   * @param int|null $archivedAgeLimit Restricting included archived groups by how long they have been archived.
+   *  Groups archived before $archivedAgeLimit days (or more) are not included in the result.
+   */
+  public function findFiltered(User $user = null, string $instanceId = null, string $search = null,
+    bool $archived = false, bool $onlyArchived = false, int $archivedAgeLimit = null)
+  {
+    $qb = $this->createQueryBuilder('g'); // takes care of softdelete cases
+
+    // Filter by instance ID...
+    $instanceId = trim($instanceId);
+    if ($instanceId) {
+      $qb->andWhere(':instanceId = g.instance')->setParameter('instanceId', $instanceId);
+    }
+
+    // Filter by user membership...
+    if ($user) {
+      $sub = $qb->getEntityManager()->createQueryBuilder()->select("gm")->from(GroupMembership::class, "gm");
+      $sub->andWhere($sub->expr()->eq('gm.user', $sub->expr()->literal($user->getId())));
+      $sub->andWhere($sub->expr()->eq('gm.status', $sub->expr()->literal(GroupMembership::STATUS_ACTIVE)));
+      $sub->andWhere($sub->expr()->eq('gm.group', 'g'));
+      $qb->andWhere($qb->expr()->exists($sub->getDQL()));
+    }
+
+    // Filter by search string...
+    if ($search) {
+      $paginationDbHelper = new PaginationDbHelper([], [ 'name' ], LocalizedGroup::class);
+      $paginationDbHelper->applySearchFilter($qb, $search);
+    }
+
+    // Filtering by archived flags...
+    $archived = $archived || $onlyArchived;
+    if (!$archived) {
+      $qb->andWhere($qb->expr()->isNull('g.archivationDate'));
+    } else if ($onlyArchived) {
+      $qb->andWhere($qb->expr()->isNotNull('g.archivationDate'));
+    }
+
+    if ($archived && $archivedAgeLimit) {
+      $minDate = new \DateTime();
+      $minDate->sub(\DateInterval::createFromDateString("$archivedAgeLimit days"));
+      $qb->andWhere('g.archivationDate >= :minDate')->setParameter('minDate', $minDate);
+    }
+
+    return $qb->getQuery()->getResult();
+  }
+
+  /**
+   * Gets an initial set of groups and produces a set of groups which have the
+   * original set as a subset and every group has its parent in the set as well.
+   * @param Group[] $groups Initial set of groups
+   */
+  public function groupsAncestralClosure(array $groups)
+  {
+    $res = [];
+    foreach ($groups as $group) {
+      $groupId = $group->getId();
+      if (array_key_exists($groupId, $res)) continue;
+      $res[$groupId] = $group;
+      foreach ($group->getParentGroupsIds() as $id) {
+        if (array_key_exists($id, $res)) break;
```

Comment:
Where in the name of our lord and savior are braces?

Reply:
As a convention, I do not write them in case of trivial statements (like if-break, if-return).

Reply:
Yeah... as a convention in our api, we write them everywhere